Source/Dafny/Resolver.cs

```diff
@@ -3320,145 +3323,144 @@ public class ModuleBindings
           bool isInteger = bin.Type.IsIntegerType;
           bool isBV = bin.E0.Type.IsBitVectorType;
           int width = isBV ? bin.E0.Type.AsBitVectorType.Width : 0;
-          if (e0 == null || e1 == null) return null;
           bool isString = e0 is string && e1 is string;
-          switch (bin.Op) {
-            case BinaryExpr.Opcode.Add:
+          switch (bin.ResolvedOp) {
+            case BinaryExpr.ResolvedOpcode.Add:
               if (isInteger) return (BigInteger) e0 + (BigInteger) e1;
               if (isBV) return ((BigInteger) e0 + (BigInteger) e1) & MaxBV(bin.Type);
-              if (isString) return (string) e0 + (string) e1;
               if (isReal) return (Basetypes.BigDec) e0 + (Basetypes.BigDec) e1;
               break;
-            case BinaryExpr.Opcode.Sub:
+            case BinaryExpr.ResolvedOpcode.Concat:
+              if (isString) return (string) e0 + (string) e1;
+              break;
+            case BinaryExpr.ResolvedOpcode.Sub:
               if (isInt && (BigInteger) e0 == 0) return -(BigInteger) e1; // Allow this case for newtypes
               if (isInteger) return (BigInteger) e0 - (BigInteger) e1;
               if (isBV) return ((BigInteger) e0 - (BigInteger) e1) & MaxBV(bin.Type);
               if (isReal) return (Basetypes.BigDec) e0 - (Basetypes.BigDec) e1;
               break;
-            case BinaryExpr.Opcode.Mul:
+            case BinaryExpr.ResolvedOpcode.Mul:
               if (isInteger) return (BigInteger) e0 * (BigInteger) e1;
               if (isBV) return ((BigInteger) e0 * (BigInteger) e1) & MaxBV(bin.Type);
               if (isReal) return (Basetypes.BigDec) e0 * (Basetypes.BigDec) e1;
               break;
-            case BinaryExpr.Opcode.BitwiseAnd:
+            case BinaryExpr.ResolvedOpcode.BitwiseAnd:
               if (isBV) return (BigInteger) e0 & (BigInteger) e1;
```

Comment:
This must be a bivector, right? If so, change `if (isBV)` to `Contract.Assert(isBV);`.
Also on L3351 and L3354.

Reply:
OK